### ISSUE TITLE
Hooked Windows Library Functions not executed

### ISSUE THREAD

**Issue by sebastianwalla** [angr/angr#821]
I try to hook a windows library function like this:

```
class recv(angr.SimProcedure):
    def run(self, s, buf, len, flags):
        print "Hooking in recv worked argc=%s , argv=%s" % (s,len)
        return 1

def main_function():
    path_to_binary = "./main.exe"
    proj = angr.Project(path_to_binary, load_options={'auto_load_libs': False})
    # hook into interesting library functions here
    proj.hook_symbol("recv", recv())
    state = proj.factory.entry_state()
    simgr = proj.factory.simgr(state,veritesting=True)
    simgr.run()
```
However this print out does never appear although recv calls are definitely done when symbolically executing the binary. It seems like the SimProcedure is never executed. This seems similiar to this [issue](https://github.com/angr/angr/issues/247) however I'm handling a windows binary so I can not overwrite the plt stubs. I therefore tried to change the pre-implemented function in angr/procedures/posix/recv.py, but this didn't work either, which makes sense to me, since this recv function is in another library than the one required here. When debugging the script I see that the recv stub in `project._sim_procedures  ` is replaced by the above SimProcedure.
But how can I now implement library hooks for windows?

The main.exe corresponds to this .cpp file, which is rather long but maybe I overlooked something in the .cpp file which causes the issue:

```
#define WIN32_LEAN_AND_MEAN

#include <windows.h>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <stdlib.h>
#include <stdio.h>


// Need to link with Ws2_32.lib, Mswsock.lib, and Advapi32.lib
#pragma comment (lib, "Ws2_32.lib")
#pragma comment (lib, "Mswsock.lib")
#pragma comment (lib, "AdvApi32.lib")


#define DEFAULT_BUFLEN 512
#define DEFAULT_PORT "27015"

int __cdecl main(int argc, char **argv)
{
	WSADATA wsaData;
	SOCKET ConnectSocket = INVALID_SOCKET;
	struct addrinfo *result = NULL,
		*ptr = NULL,
		hints;
	char *sendbuf = "this is a test";
	char recvbuf[DEFAULT_BUFLEN];
	int iResult;
	int recvbuflen = DEFAULT_BUFLEN;

	// Validate the parameters
	if (argc != 2) {
		printf("usage: %s server-name\n", argv[0]);
		return 1;
	}

	// Initialize Winsock
	iResult = WSAStartup(MAKEWORD(2, 2), &wsaData);
	if (iResult != 0) {
		printf("WSAStartup failed with error: %d\n", iResult);
		return 1;
	}

	ZeroMemory(&hints, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_protocol = IPPROTO_TCP;

	// Resolve the server address and port
	iResult = getaddrinfo(argv[1], DEFAULT_PORT, &hints, &result);
	if (iResult != 0) {
		printf("getaddrinfo failed with error: %d\n", iResult);
		WSACleanup();
		return 1;
	}

	// Attempt to connect to an address until one succeeds
	for (ptr = result; ptr != NULL;ptr = ptr->ai_next) {

		// Create a SOCKET for connecting to server
		ConnectSocket = socket(ptr->ai_family, ptr->ai_socktype,
			ptr->ai_protocol);
		if (ConnectSocket == INVALID_SOCKET) {
			printf("socket failed with error: %ld\n", WSAGetLastError());
			WSACleanup();
			return 1;
		}

		// Connect to server.
		iResult = connect(ConnectSocket, ptr->ai_addr, (int)ptr->ai_addrlen);
		if (iResult == SOCKET_ERROR) {
			closesocket(ConnectSocket);
			ConnectSocket = INVALID_SOCKET;
			continue;
		}
		break;
	}

	freeaddrinfo(result);

	if (ConnectSocket == INVALID_SOCKET) {
		printf("Unable to connect to server!\n");
		WSACleanup();
		return 1;
	}

	// Send an initial buffer
	iResult = send(ConnectSocket, sendbuf, (int)strlen(sendbuf), 0);
	if (iResult == SOCKET_ERROR) {
		printf("send failed with error: %d\n", WSAGetLastError());
		closesocket(ConnectSocket);
		WSACleanup();
		return 1;
	}

	printf("Bytes Sent: %ld\n", iResult);

	// shutdown the connection since no more data will be sent
	iResult = shutdown(ConnectSocket, SD_SEND);
	if (iResult == SOCKET_ERROR) {
		printf("shutdown failed with error: %d\n", WSAGetLastError());
		closesocket(ConnectSocket);
		WSACleanup();
		return 1;
	}

	FD_SET writeSet;
	FD_SET readSet;

	FD_SET(ConnectSocket, &readSet);
	FD_SET(ConnectSocket, &writeSet);
	// nfds is ignored
	if (select(1, &readSet, &writeSet, NULL, NULL) > 0) {
		iResult = recv(ConnectSocket, recvbuf, recvbuflen, 0);
	}
	// Receive until the peer closes the connection
	do {

			iResult = recv(ConnectSocket, recvbuf, recvbuflen, 0);
			if (iResult > 0)
				printf("Bytes received: %d\n", iResult);
			else if (iResult == 0)
				printf("Connection closed\n");
			else
				printf("recv failed with error: %d\n", WSAGetLastError());
			iResult = recv(ConnectSocket, recvbuf, recvbuflen, 0);

	} while (iResult > 0);
	

	// cleanup
	closesocket(ConnectSocket);
	WSACleanup();

	return 0;
}
  
```
  
Note that according to the [Programming SimProcedures tutorial](https://docs.angr.io/docs/simprocedures.html) I should pass rather `recv` as a second argument to the `hook_symbol()` function than `recv()` . However then one will get a warning that passing `recv` is deprecated. I of course tried both and none of those two versions got me printed the expected _Hooking in recv worked..._

**Comment by rhelmot** [comment]
Hello! First, thank you for pointing out the inconsistency in the docs - I've updated it to fix the deprecated code.

Your issue has nothing to do with the way you're hooking, and everything to do with angr's poor support for windows programs - or rather, the serious issues with trying to do blind binary analysis on files compiled with msvc. At least when I compile and run this code, angr spins off into oblivion with a state explosion trying to execute the initializer functions - we don't start execution from main, we start from the entry point. Between the amount of code that's statically linked into the executable (as opposed to linux's standard `__libc_start_main`) and the horrifying reliance on dynamic loading  (api-ms-win-core-synch-l1-2-0.dll anyone?), it's a nightmare even getting to main.

Additionally, there is the serious issue with the stdcall calling convention, and when we don't know how many arguments a given function is supposed to take, and the real implementation of the function isn't available to run, the stack will be unbalanced and emulation will fail. @schieb was working on a library of common functions and their arguments to add to angr to solve this problem, but I haven't heard from him recently.

Long story short: you're going to need to hold angr's hand a lot more aggressively if you want any sort of analysis like this to work. This will involve one or both of implementing simprocedure stubs for problematic library functions and providing more specific analysis goals than "symbolically execute from the entry point."

You can debug things like this by either stepping until you see a symbolic branch (`simgr.step(until=lambda lsm: len(lsm.active) != 1)`) or just ^Cing the emulation and examining what's in the simgr at any given point.

**Comment by sebastianwalla** [comment]
With the veritesting attribute set to true it runs for me, using `simgr.run()`, in under a minute, but I should have mentioned, that in the actual code I use the explore function with the find function forcing the execution to follow a predefined path. Without veritesting set to true, it took me about ten minutes till this path was finished. However this path involves the basic blocks calling the recv function but the corresponding hooks are never called. 

I will definitely try going through this step by step again but the addresses of elements on the active stash also matched those of the basic blocks containing the recv calls. But better checking twice than being sorry afterwards.


> Additionally, there is the serious issue with the stdcall calling convention, and when we don't know how many arguments a given function is supposed to take, and the real implementation of the function isn't available to run, the stack will be unbalanced and emulation will fail. @schieb was working on a library of common functions and their arguments to add to angr to solve this problem, but I haven't heard from him recently.

Okay so the best would be having this libraries also loaded but overwriting the most common functions with hooks to speed things up. Would it be necessary to write a extra calling convention for each of this hooks or will angr extract this, since it probably assumes the stdcall calling convention (for windows) and from the hooks it can infer the number of arguments?

But in order to overwrite the most common used functions I again need to get the hooking working. Or do you think that the stack simply got unbalanced and this caused that the hooks did not got called although the basic blocks which should trigger calls to those hooks was symbolically executed?

**Comment by sebastianwalla** [comment]
I debugged it now step by step and found out what was the problem: I assumed that the explore function would only return successfully if it found a state matching the find function and otherwise throw an exception. But that is not the case. So I falsely assumed that the basic blocks would have been covered, but instead the explore function just returned immediately since it got not active stashes but only an errored one back. Therefore the basic blocks containing the hooks have never been executed. I'm sorry it was a wrong assumption on my side. As explained the `__security_init_cookie(void)` function already fails for me so next I will try to overwrite this with a hook and then look further step by step.

**Comment by schieb** [comment]
> Additionally, there is the serious issue with the stdcall calling convention, and when we don't know how many arguments a given function is supposed to take, and the real implementation of the function isn't available to run, the stack will be unbalanced and emulation will fail. @schieb was working on a library of common functions and their arguments to add to angr to solve this problem, but I haven't heard from him recently.

This was a while back, but iirc, I used @rhelmot's [callee cleanup finder](https://github.com/angr/angr/blob/master/angr/analyses/callee_cleanup_finder.py) on kernel32 and it classified all but one function correctly. Next week I will have time to look into the failure, test advapi, and submit a PR for review.

> Okay so the best would be having this libraries also loaded but overwriting the most common functions with hooks to speed things up. Would it be necessary to write a extra calling convention for each of this hooks or will angr extract this, since it probably assumes the stdcall calling convention (for windows) and from the hooks it can infer the number of arguments?

Apart from using the above mentioned callee cleanup finder, there is no way for angr to automatically determine the correct calling convention when stubbing external functions with simprocedures. Since each library defines its own calling convention, it is not correct for angr to assume stdcall for all windows x86 dlls. Instead, each [simlibrary](https://github.com/angr/angr/blob/master/angr/procedures/definitions/__init__.py) can specify the calling convention for the corresponding dll that it implements [[example](https://github.com/angr/angr/blob/bef5cb727f7d88c8513f73b7a4b861ce4ae340ea/angr/procedures/definitions/kernel32.py#L8)]. As long as your simprocedure has the correct number of arguments and is defined as callee-cleanup, angr will effectively emulate a `ret X` upon completion of the simprocedure, where X corresponds to the number of arguments. See [here](https://github.com/angr/angr/blob/922348c4989d66b437a6b710f9ce2f9ce2d4aec0/angr/calling_conventions.py#L537-L569) for the actual implementation.
